romancal/saturation/tests/test_saturation.py

```diff
@@ -0,0 +1,274 @@
+
+    # Make sure all groups are flagged
+    assert np.all(output.groupdq[:, 5, 5] == dqflags.group['SATURATED'])
+
+def test_dq_propagation(setup_wfi_datamodels):
+    '''Check PIXELDQ propagation.'''
+
+    # Create inputs, data, and saturation maps
+    ngroups = 5
+    nrows = 20
+    ncols = 20
+    dqval1 = 5
+    dqval2 = 10
+
+    data, satmap = setup_wfi_datamodels(ngroups, nrows, ncols)
+
+    # Add DQ values to the data and reference file
+    data.pixeldq[5, 5] = dqval1
+    satmap.dq[5, 5] = dqval2
+
+    # Run the pipeline
+    output = flag_saturation(data, satmap)
+
+    # Make sure DQ values from data and reference file are added in the output
+    assert output.pixeldq[5, 5] == dqval1 + dqval2
+
+
+def test_no_sat_check(setup_wfi_datamodels):
+    '''Check that pixels flagged with NO_SAT_CHECK in the reference file get
+       added to the DQ mask and are not flagged as saturated.'''
+
+    # Create inputs, and data and saturation models
+    ngroups = 5
+    nrows = 20
+    ncols = 20
+    satvalue = 60000
+
+    data, satmap = setup_wfi_datamodels(ngroups, nrows, ncols)
+
+    # Add ramp values up to the saturation limit
+    data.data[0, 5, 5] = 10
+    data.data[1, 5, 5] = 20000
+    data.data[2, 5, 5] = 40000
+    data.data[3, 5, 5] = 60000
+    data.data[4, 5, 5] = 62000   # Signal reaches saturation limit
+
+    # Set saturation value in the saturation model & DQ value for NO_SAT_CHECK
+    satmap.data[5, 5] = satvalue
+    satmap.dq[5, 5] = dqflags.pixel['NO_SAT_CHECK']
+
+    # Also set an existing DQ flag in input science data
+    data.pixeldq[5, 5] = dqflags.pixel['DO_NOT_USE']
+
+    # Run the pipeline
+    output = flag_saturation(data, satmap)
+
+    # Make sure output GROUPDQ does not get flagged as saturated
+    # Make sure PIXELDQ is set to NO_SAT_CHECK and original flag
+    assert np.all(output.groupdq[:, 5, 5] != dqflags.group['SATURATED'])
+    assert output.pixeldq[5, 5] == (dqflags.pixel['NO_SAT_CHECK'] +
+                                    dqflags.pixel['DO_NOT_USE'])
+
+
+def test_nans_in_mask(setup_wfi_datamodels):
+    '''Check that pixels in the reference files that have value NaN are not
+       flagged as saturated in the data and that in the PIXELDQ array the
+       pixel is set to NO_SAT_CHECK.'''
+
+    # Create inputs, and data and saturation models
+    ngroups = 5
+    nrows = 20
+    ncols = 20
+
+    data, satmap = setup_wfi_datamodels(ngroups, nrows, ncols)
+
+    # Add ramp values up to the saturation limit
+    data.data[0, 5, 5] = 10
+    data.data[1, 5, 5] = 20000
+    data.data[2, 5, 5] = 40000
+    data.data[3, 5, 5] = 60000
+    data.data[4, 5, 5] = 62000
+
+    # Set saturation value for pixel to NaN
+    satmap.data[5, 5] = np.nan
+
+    # Run the pipeline
+    output = flag_saturation(data, satmap)
+
+    # Check that output GROUPDQ is not flagged as saturated
+    assert np.all(output.groupdq[:, 5, 5] != dqflags.group['SATURATED'])
```

Comment:
Just curious - what would the groupdq be set to (if anything) in this case ? (I suppose it doesn't matter since pdq is set to 'NO_SAT_CHECK' )

Reply:
Nothing? The values set here are below the threshold, so they are all valid.